src/NetworkRouteSelector_P.cpp

```diff
@@ -0,0 +1,734 @@
+}
+
+int OsmAnd::NetworkRouteSelector_P::connectSimpleMerge(QMap<int64_t, QList<NetworkRouteSegmentChain>> & chains,
+                       QMap<int64_t, QList<NetworkRouteSegmentChain>> & endChains, int rad, int radE) const
+{
+    int merged = 1;
+    while (merged > 0 /*&& !isCancelled()*/)
+    {
+        int rs = reverseToConnectMore(chains, endChains, rad, radE);
+        merged = connectSimpleStraight(chains, endChains, rad, radE);
+        OsmAnd::LogPrintf(LogSeverityLevel::Info, "Simple merged: %d, reversed: %d (radius %d %d)", merged, rs, rad, radE);
+    }
+    return merged;
+}
+
+int OsmAnd::NetworkRouteSelector_P::connectSimpleStraight(QMap<int64_t, QList<NetworkRouteSegmentChain>> & chains,
+                          QMap<int64_t, QList<NetworkRouteSegmentChain>> & endChains, int rad, int radE) const
+{
+    int merged = 0;
+    bool changed = true;
+    while (changed /*&& !isCancelled()*/)
+    {
+        changed = false;
+        QList<int64_t> longPoints(chains.keys());//copy of chains keys
+        for (int64_t startPnt : longPoints)
+        {
+            if (changed)
+            {
+                break;
+            }
+            auto iterator = chains.find(startPnt);
+            auto & lst = iterator.value();
+            for (NetworkRouteSegmentChain & it : lst)
+            {
+                auto endPoint = it.getEndPoint();
+                int64_t pnt = owner->rCtx->convertPointToLong(endPoint.x, endPoint.y);
+                QList<NetworkRouteSegmentChain> connectNextLst = getByPoint(chains, pnt, radE, &it);
+                connectNextLst = filterChains(connectNextLst, it, rad, true);
+                QList<NetworkRouteSegmentChain> connectToEndLst = getByPoint(endChains, pnt, radE, &it);
+                connectToEndLst = filterChains(connectToEndLst, it, rad, false);
+                if (connectToEndLst.size() > 0)
+                {
+                    for (auto & next : connectNextLst)
+                    {
+                        connectToEndLst.removeAll(next);
+                    }
+                }
+                // no alternative join
+                if (connectNextLst.size() == 1 && connectToEndLst.size() == 0)
+                {
+                    NetworkRouteSegmentChain toAdd = connectNextLst.at(0);
+                    chainAdd(chains, endChains, it, toAdd);
+                    changed = true;
+                    merged++;
+                    break;
+                }
+            }
+        }
+    }
+    return merged;
+}
+
+int OsmAnd::NetworkRouteSelector_P::reverseToConnectMore(QMap<int64_t, QList<NetworkRouteSegmentChain>> & chains,
+                         QMap<int64_t, QList<NetworkRouteSegmentChain>> & endChains, int rad, int radE) const
+{
+    int reversed = 0;
+    const QList<int64_t> & longPoints = chains.keys();
+    for (const int64_t & startPnt : longPoints)
+    {
+        auto it = chains.find(startPnt);
+        QList<NetworkRouteSegmentChain> vls = it.value();
```

Comment:
To iterate map entries use `rangeOf(constOf(map))` !

Reply:
Changed to
//chains.values() - copy only, chains is changing inside
for (QList<NetworkRouteSegmentChain> & vls : chains.values())